### DISCUSSION TITLE
MidiMapping in z_config

### DISCUSSION THREAD

**Discussion by MichaelPNolan** [marcel-licence/esp32_basic_synth#47] [Q&A]
So - I've integrated USB_Host_Midi module into this synth - I still haven't put a DC or the ADC but I'm just looking to understand further integration. Can someone help me with this:

This was here already
```
struct midiMapping_s midiMapping =
{
    Synth_NoteOn,
    Synth_NoteOff,
    Synth_PitchBend,
    Synth_ModulationWheel,
    edirolMapping,
    sizeof(edirolMapping) / sizeof(edirolMapping[0]),
};

```
There is a uniformity of some parts of the various projects and variations/versions of config.h and z_config

These struct definitions - I've brought over - these seem to be pointers to function calls and this is a little beyond my limited coding experience - at least in recent years but I do have a formal education in Computer Science so nothing to do but work at it step by step. My guess (and sure, I can experiment and test) is that I can replace NULL to make the midi events get mapped to call Synth to add notes to it's structure of notes (voices) which it then processes at the sample rate called from the main loop.

Assumedly I can discover the notes/channel of my Launchpad and make a mapping something like the edirolMapping so I can trigger notes from that for testing and fun. So why is Midi_SendShortMessage. What else might I have to change in the usbMidiHost.ino to process this mapping?

```
struct usbMidiMappingEntry_s usbMidiMappingEntries[] =
{
    {
        NULL,
        Midi_SendShortMessage,
        NULL,
        NULL,
        0x01,
    },
};

struct usbMidiMapping_s usbMidiMapping =
{
    NULL,
    NULL,
    usbMidiMappingEntries,
    sizeof(usbMidiMappingEntries) / sizeof(usbMidiMappingEntries[0]),
};
```

**Comment by MichaelPNolan** [comment]
I think I understand it a bit better, no worries if people don't have time to answer. You kind of have to define the function types in the library type before you can map them to functions. So I'm just copy/pasting the code to play note from midi_interface.ino into usbMidiHost.ino where it has short message handled.

**Reply by marcel-licence** [reply]
If you look into usbMidiHost.ino you will find the definition of struct usbMidiMappingEntry_s. My idea was to allow creating multiple entries to connect the different usb midi lines to different midi message handler functions.

The second entry "shortMsg" will be called if a short message has been received. So you can redirect the messages into the application. I see I have a local modification not commited yet. You attempt is correct.

**Reply by MichaelPNolan** [reply]
If you were going to make a commit before I try to work with it that could be good. Otherwise I guess I'll find new code. I'll try to replicate something with handler functions as you say.
